### ISSUE TITLE
Permission issues with DorisMonitor

### ISSUE THREAD

**Issue by swapkh91** [linsoss/doris-operator#19]
@Al-assad 
while deploying DorisMonitor using yaml provided [here](https://github.com/linsoss/doris-operator/blob/dev/examples/advanced/doris-monitor.yaml), getting permission errors on directories, online solutions are mainly on giving securityContext, initContainers etc which can change the folder permissions during creation

grafana:
```
Error: ✗ failed to connect to database: failed to create SQLite database file "/data/grafana/grafana.db": open /data/grafana/grafana.db: permission denied

GF_PATHS_DATA='/data/grafana' is not writable.

You may have issues with file permissions, more information here: http://docs.grafana.org/installation/docker/#migrate-to-v51-or-later
```

prometheus:
```
ts=2023-11-10T10:49:46.440Z caller=query_logger.go:91 level=error component=activeQueryTracker msg="Error opening query log file" file=/data/prometheus/queries.active err="open /data/prometheus/queries.active: permission denied"
```

loki:
```
level=error ts=2023-11-10T11:09:43.453704809Z caller=log.go:230 msg="error running loki" err="mkdir /data/loki: permission denied\nerror creating object client\ngithub.com/grafana/loki/pkg/storage.
```

promtail was created successfully

**Comment by Al-assad** [comment]
@swapkh91 Can you post the descriptions of the deployments for Grafana, Prometheus and Loki? It seems to be an issue with PVC mounting.
Just like: `kubectl describe deployment basic-monitor-grafana`
<img width="1177" alt="image" src="https://github.com/linsoss/doris-operator/assets/22493821/2e70ba75-f1aa-42ea-8f1a-4aadbf03022b">

**Comment by swapkh91** [comment]
@Al-assad

grafana
```
Name:                   basic-monitor-grafana
Namespace:              doris
CreationTimestamp:      Tue, 14 Nov 2023 11:19:30 +0530
Labels:                 app.kubernetes.io/component=grafana
                        app.kubernetes.io/instance=basic
                        app.kubernetes.io/managed-by=doris-operator
                        app.kubernetes.io/name=doris-cluster
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               app.kubernetes.io/component=grafana,app.kubernetes.io/instance=basic,app.kubernetes.io/managed-by=doris-operator,app.kubernetes.io/name=doris-cluster
Replicas:               1 desired | 1 updated | 1 total | 0 available | 1 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:       app.kubernetes.io/component=grafana
                app.kubernetes.io/instance=basic
                app.kubernetes.io/managed-by=doris-operator
                app.kubernetes.io/name=doris-cluster
  Annotations:  al-assad.github.io/grafana-conf: a261f696b38a2c4c59e07a094dd5e12b
  Containers:
   grafana:
    Image:      grafana/grafana:9.5.2
    Port:       3000/TCP
    Host Port:  0/TCP
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:      250m
      memory:   500Mi
    Liveness:   http-get http://:3000/api/health delay=0s timeout=5s period=10s #success=1 #failure=3
    Readiness:  http-get http://:3000/api/health delay=0s timeout=5s period=10s #success=1 #failure=3
    Environment:
      GF_PATHS_DATA:               /data/grafana
      GF_SECURITY_ADMIN_USER:      <set to the key 'user' in secret 'basic-monitor-grafana-config'>      Optional: false
      GF_SECURITY_ADMIN_PASSWORD:  <set to the key 'password' in secret 'basic-monitor-grafana-config'>  Optional: false
    Mounts:
      /data/grafana from grafana-data (rw)
      /etc/grafana/provisioning/dashboards from grafana-dashboard (rw)
      /etc/grafana/provisioning/datasources from grafana-datasource (rw)
  Volumes:
   grafana-datasource:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      basic-monitor-grafana-secret
    Optional:  false
   grafana-dashboard:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      basic-monitor-grafana-secret
    Optional:  false
   grafana-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  basic-monitor-grafana-pvc
    ReadOnly:   false
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      False   MinimumReplicasUnavailable
  Progressing    True    ReplicaSetUpdated
OldReplicaSets:  <none>
NewReplicaSet:   basic-monitor-grafana-9874c5b4b (1/1 replicas created)
Events:
  Type    Reason             Age    From                   Message
  ----    ------             ----   ----                   -------
  Normal  ScalingReplicaSet  7m43s  deployment-controller  Scaled up replica set basic-monitor-grafana-9874c5b4b to 1
```

loki
```
Name:                   basic-monitor-loki
Namespace:              doris
CreationTimestamp:      Tue, 14 Nov 2023 11:19:30 +0530
Labels:                 app.kubernetes.io/component=loki
                        app.kubernetes.io/instance=basic
                        app.kubernetes.io/managed-by=doris-operator
                        app.kubernetes.io/name=doris-cluster
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               app.kubernetes.io/component=loki,app.kubernetes.io/instance=basic,app.kubernetes.io/managed-by=doris-operator,app.kubernetes.io/name=doris-cluster
Replicas:               1 desired | 1 updated | 1 total | 0 available | 1 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:       app.kubernetes.io/component=loki
                app.kubernetes.io/instance=basic
                app.kubernetes.io/managed-by=doris-operator
                app.kubernetes.io/name=doris-cluster
  Annotations:  al-assad.github.io/loki-conf: cf9026ab949b1b0e84bbb86af41f6574
  Containers:
   loki:
    Image:       grafana/loki:2.9.1
    Ports:       3100/TCP, 9095/TCP
    Host Ports:  0/TCP, 0/TCP
    Args:
      -config.file=/etc/loki/loki.yml
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:        500m
      memory:     500Mi
    Liveness:     http-get http://:3100/ready delay=0s timeout=1s period=10s #success=1 #failure=3
    Readiness:    http-get http://:3100/ready delay=45s timeout=1s period=10s #success=1 #failure=3
    Environment:  <none>
    Mounts:
      /data from data (rw)
      /etc/loki from config (rw)
  Volumes:
   config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      basic-monitor-loki-config
    Optional:  false
   data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  basic-monitor-loki-pvc
    ReadOnly:   false
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      False   MinimumReplicasUnavailable
  Progressing    True    ReplicaSetUpdated
OldReplicaSets:  <none>
NewReplicaSet:   basic-monitor-loki-d79fdbb7b (1/1 replicas created)
Events:
  Type    Reason             Age    From                   Message
  ----    ------             ----   ----                   -------
  Normal  ScalingReplicaSet  9m28s  deployment-controller  Scaled up replica set basic-monitor-loki-d79fdbb7b to 1
```

prometheus
```
Name:                   basic-monitor-prometheus
Namespace:              doris
CreationTimestamp:      Tue, 14 Nov 2023 11:19:30 +0530
Labels:                 app.kubernetes.io/component=prometheus
                        app.kubernetes.io/instance=basic
                        app.kubernetes.io/managed-by=doris-operator
                        app.kubernetes.io/name=doris-cluster
Annotations:            deployment.kubernetes.io/revision: 1
Selector:               app.kubernetes.io/component=prometheus,app.kubernetes.io/instance=basic,app.kubernetes.io/managed-by=doris-operator,app.kubernetes.io/name=doris-cluster
Replicas:               1 desired | 1 updated | 1 total | 0 available | 1 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app.kubernetes.io/component=prometheus
                    app.kubernetes.io/instance=basic
                    app.kubernetes.io/managed-by=doris-operator
                    app.kubernetes.io/name=doris-cluster
  Annotations:      al-assad.github.io/prometheus-conf: 01d5ed9382711a9a5d64059ff687915d
  Service Account:  doris-monitor
  Containers:
   prometheus:
    Image:      prom/prometheus:v2.37.8
    Port:       9090/TCP
    Host Port:  0/TCP
    Args:
      --config.file=/etc/prometheus/prometheus.yml
      --storage.tsdb.path=/data/prometheus
      --web.enable-lifecycle
      --storage.tsdb.retention.time=30d
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:        500m
      memory:     500Mi
    Liveness:     http-get http://:9090/-/ready delay=0s timeout=1s period=5s #success=1 #failure=30
    Readiness:    http-get http://:9090/-/ready delay=0s timeout=1s period=5s #success=1 #failure=120
    Environment:  <none>
    Mounts:
      /data/prometheus from prometheus-data (rw)
      /etc/prometheus from prometheus-config (rw)
  Volumes:
   prometheus-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      basic-monitor-promtheus-config
    Optional:  false
   prometheus-data:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  basic-monitor-prometheus-pvc
    ReadOnly:   false
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      False   MinimumReplicasUnavailable
  Progressing    False   ProgressDeadlineExceeded
OldReplicaSets:  <none>
NewReplicaSet:   basic-monitor-prometheus-859fbbc67d (1/1 replicas created)
Events:
  Type    Reason             Age   From                   Message
  ----    ------             ----  ----                   -------
  Normal  ScalingReplicaSet  10m   deployment-controller  Scaled up replica set basic-monitor-prometheus-859fbbc67d to 1
```

**Comment by swapkh91** [comment]
@Al-assad did you get a chance to look at this?

**Comment by Al-assad** [comment]
> @Al-assad did you get a chance to look at this?你有机会看这个吗？

Sure, I'm a little busy this week and I'll work on it over the weekend.

**Comment by swapkh91** [comment]
@Al-assad I was able to manually deploy grafana and prometheus

grafana.yaml
```
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: grafana-storage
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: grafana
data:
  allow-snippet-annotations: "false"
  grafana.ini: |
    [analytics]
    check_for_updates = true
    [grafana_net]
    url = https://grafana.net
    [log]
    mode = console
    [paths]
    data = /var/lib/grafana/
    logs = /var/log/grafana
    plugins = /var/lib/grafana/plugins
    provisioning = /etc/grafana/provisioning

  datasources.yaml: |
    apiVersion: 1
    datasources:
    - access: proxy
      editable: true
      name: doris-cluster
      type: prometheus
      url: http://prometheus:9090
      version: 1
  dashboardproviders.yaml: |
    apiVersion: 1
    providers:
    - disableDeletion: false
      folder: ''
      name: Default
      options:
        path: /etc/grafana/provisioning/dashboards
      orgId: 1
      type: file

---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: grafana
  name: grafana
spec:
  replicas: 1
  selector:
    matchLabels:
      app: grafana
  template:
    metadata:
      labels:
        app: grafana
    spec:
      nodeSelector:
        cloud.google.com/gke-nodepool: doris-monitor-pool
      containers:
      - image: grafana/grafana:10.2.1
        name: grafana
        ports:
        - containerPort: 3000
          name: http
        volumeMounts:
          - name: grafana-storage
            mountPath: /var/lib/grafana
        resources:
          requests:
            cpu: 250m
            memory: 500Mi
          limits:
            cpu: 300m
            memory: 750Mi
      volumes:
        - name: grafana-storage
          persistentVolumeClaim:
            claimName: grafana-storage
      securityContext:
        runAsNonRoot: true
        runAsUser: 65534
        fsGroup: 472

---
apiVersion: v1
kind: Service
metadata:
  name: grafana
spec:
  type: ClusterIP
  ports:
    - name: grafana-port
      port: 3000
      protocol: TCP
      targetPort: 3000
  selector:
    app: grafana
```

prometheus yaml is huge so not attaching here

I think what works is
```
securityContext:
        runAsNonRoot: true
        runAsUser: 65534
        fsGroup: 472
```
and for prometheus
```
securityContext:
        fsGroup: 65534
        runAsGroup: 65534
        runAsNonRoot: true
        runAsUser: 65534
```

I then imported the dashboard json from [here](https://github.com/linsoss/doris-operator/blob/dev/internal/template/monitor/grafana-dashboards.json) and changed the prometheus uid from my datasource and was able to see the dashboard

![image](https://github.com/linsoss/doris-operator/assets/5755405/6b538471-b941-45c8-a262-c90dd9209052)

**Comment by Al-assad** [comment]
@swapkh91 Hi, I think a better way would be to refactor prometheus, grafana, loki resources into statefulset, just like promethues-operator's does. I will complete it soon.

**Comment by Al-assad** [comment]
@swapkh91 Hi Khare, I just released a version [1.0.2](https://github.com/linsoss/doris-operator/releases/tag/v1.0.2) that may solve your issue!
